Homework for sprint №21 (DB, Room)

sobolevkir opened the pull request:
### Реализован экран "Избранные треки" с использованием БД и библиотеки Room.

В задании было сказано, что в интеракторе **FavoritesInteractor** нам нужны только  3 метода:
-добавления трека в избранное
-удаления трека из избранного
-получения списка треков из избранного,
а также ещё один метод реализован в самом DAO-интерфейсе для получения списка Id избранных треков.

Если оставить всё так, то тогда был бы баг в таком кейсе:
1. пользователь находит трек
2. далее нажимает на него и попадает на экран плеера
3. нажимает на кнопку добавления в избранное: трек добавляется, кнопка активна
4. выходит из активити плеера по нажатию кнопки назад, активити закрывается и пользователь возвращается обратно на экран поиска
5. пользователь снова нажимает на тот же трек
6. при открытии плеера видно, что кнопка не активна (хотя трек попал в таблицу)

Этот баг происходит потому, что передача данных в плеер осуществляется через аргумент в навигации.
Поэтому мне пришлось добавить ещё один метод в **FavoritesInteractor** - _isTrackFavorite(trackId: String)_
и сделать проверку в PlayerViewModel на то, является ли трек избранным. Так всё работает, но..

В связи с этим возникает вопрос, для чего тогда нужно проверять, является ли трек избранным на этапе поиска, если в плеер при повторном открытии трека всё-равно передаётся трек в виде Serialized-аргумента со старыми данными?
Можно ли отказаться от проверок на вхождение трека в избранное на этапе поиска и передавать в RecyclerView треки без этого параметра, а проверять уже в самом плеере?
Или лучше отказаться от проверки на экране плеера, при этом апдейтить экран поиска (при переходе на него с экрана плеера по нажатию back), отправляя новый поисковый запрос при возвращении на экран поиска? (тогда нужно будет это делать не только в поиске, но и в истории, и на самом экране избранных треков, и в плейлистах..

Как лучше реализовать эту проверку красиво и эстетично?

Заранее спасибо за ответ!

sobolevkir committed:
Реализован экран "Избранные треки" с использованием БД и библиотеки Room.

DmitrijEgorow commented:
В начальный момент времени при показе кнопки Play и после окончания проигрывания не показывается длительность отрывка согласно дизайн-макету


https://github.com/user-attachments/assets/8120ffdb-d86e-4fdd-bd93-01f22ca3c304

DmitrijEgorow commented on app/src/main/java/com/sobolevkir/playlistmaker/common/data/db/dao/FavoriteTrackDao.kt:
```diff
@@ -0,0 +1,25 @@
+    suspend fun deleteTrack(track: FavoriteTrackEntity)
+
+    @Query("SELECT * FROM favorite_tracks_table")
+    suspend fun getTracks(): List<FavoriteTrackEntity>
```
Room позволяет возвращать сразу Flow<List<.>>

sobolevkir replied:
сделано! ✅

DmitrijEgorow commented on app/src/main/java/com/sobolevkir/playlistmaker/common/data/db/entity/FavoriteTrackEntity.kt:
```diff
@@ -0,0 +1,21 @@
+@Entity(tableName = "favorite_tracks_table")
+data class FavoriteTrackEntity(
+    @PrimaryKey @ColumnInfo(name = "track_id")
+    val trackId: String = "-1",
```
Если это UUID, то я соглашусь с типом String
А иначе почему не Long?

sobolevkir replied:
Поменял везде на Long ✅

DmitrijEgorow commented:
> Этот баг происходит потому, что передача данных в плеер осуществляется через аргумент в навигации. 


Именно поэтому мне лично не нравится такой способ передачи данных. Дополнительно, большие массивы данных так не передашь.

А теперь представим, что мы расширили функционал приложения, и стало возможным управлять воспроизведением текущей композиции (и осуществлять простановку лайка) и через открытую активность PlayerActivity, и одновременно через уведомление (в шторке), и всё это должно синхронизироваться. Простановка лайка через уведомление должна бы отразиться при разворачивании активности плеера. При передачи через навигацию такую доработку сделать сложнее, но в текущей реализации это допустимо.

Room позволяет возвращать сразу Flow. Как только данные в БД изменяются, Room генерирует новые данные, которые передаются через Flow. Если так реализовать, то можно будет через Database Inspector изменять данные в БД и тут же смотреть обновления интерфейса.

DmitrijEgorow commented:
>  Можно ли отказаться от проверок на вхождение трека в избранное на этапе поиска и передавать в RecyclerView треки без этого параметра, а проверять уже в самом плеере? 

В принципе, можно. Можно дозапрашивать через getTracksIds(), проверять уже в плеере, и позже окрашивать кнопку лайка. Часть приложений так работает

DmitrijEgorow commented:
>  при этом апдейтить экран поиска (при переходе на него с экрана плеера по нажатию back), отправляя новый поисковый запрос при возвращении на экран поиска? 

не стоит

sobolevkir commented:
> > Этот баг происходит потому, что передача данных в плеер осуществляется через аргумент в навигации.
> 
> Именно поэтому мне лично не нравится такой способ передачи данных. Дополнительно, большие массивы данных так не передашь.
> 
> А теперь представим, что мы расширили функционал приложения, и стало возможным управлять воспроизведением текущей композиции (и осуществлять простановку лайка) и через открытую активность PlayerActivity, и одновременно через уведомление (в шторке), и всё это должно синхронизироваться. Простановка лайка через уведомление должна бы отразиться при разворачивании активности плеера. При передачи через навигацию такую доработку сделать сложнее, но в текущей реализации это допустимо.
> 
> Room позволяет возвращать сразу Flow. Как только данные в БД изменяются, Room генерирует новые данные, которые передаются через Flow. Если так реализовать, то можно будет через Database Inspector изменять данные в БД и тут же смотреть обновления интерфейса.

Имеешь в виду, что трек перед открытием плеера нужно сохранять в БД и в активности плеера брать его из БД?
Как лучше всего передать данные о треке в активность плеера в нашем приложении?
И что такое Database Inspector и как его реализовать в данном случае? Можно чуть подробнее, пожалуйста? 🙏

sobolevkir commented:
> ```
> В начальный момент времени при показе кнопки Play и после окончания проигрывания не показывается длительность отрывка согласно дизайн-макету
> ```
> 
>  Screen_recording_20240726_232515.mp4

исправлено ✅

DmitrijEgorow commented:
>  И что такое Database Inspector и как его реализовать в данном случае? Можно чуть подробнее, пожалуйста? 🙏

Внизу-слева есть раздел App Inspection. Если будут трудности, спроси в чате. Будет работать примерно так


https://github.com/user-attachments/assets/f9295734-b733-4143-b9a9-77dc9114ceff

DmitrijEgorow commented:
> Имеешь в виду, что трек перед открытием плеера нужно сохранять в БД и в активности плеера брать его из БД? Как лучше всего передать данные о треке в активность плеера в нашем приложении? И что такое Database Inspector и как его реализовать в данном случае? Можно чуть подробнее, пожалуйста? 🙏

Один текущий трек в БД – overwhelm. Самое простое – передавать через навигацию, как сейчас
А если бы оба были фрагментами, то [https://developer.android.com/guide/fragments/communicate#fragments](https://developer.android.com/guide/fragments/communicate#fragments)

sobolevkir commented:
> > И что такое Database Inspector и как его реализовать в данном случае? Можно чуть подробнее, пожалуйста? 🙏
> 
> Внизу-слева есть раздел App Inspection. Если будут трудности, спроси в чате. Будет работать примерно так
> 
>  2024-08-02.23.22.09.mp4

Да, всё так и работает теперь!
Но всё же я не понимаю, как мне сделать обновление списка в результатах поиска и в истории через этот flow, ведь треки, которые в результатах поиска находятся не из базы данных берутся, а через flow с сервера и поэтому этот список не обновляется при попадании трека в избранные.
Может ты имел в виду, что нужно сделать апдейт списка (без сетевого запроса, но через интерактор) в методе onResume фрагмента?
Мне нужно понимание, какое решение здесь наиболее удачным будет для обновления списка в результатах поиска и в истории...

sobolevkir committed:
Доработана работа c избранным:
Теперь на экране плеера статус трека в избранном дополнительно не запрашивается. Вместо этого используется синхронизация и обновление списков в результатах поиска и истории через Flow из БД. Теперь в будущем можно будет реализовать возможность просмотра статуса избранного прямо в списке (например, сделать специальную иконку в элементах списков).

Кроме этого обратный порядок вывода списка избранных треков на экране Медиатеки теперь релизован с помощью дополнительного поля timestamp.

Также внесено исправление на экране плеера: теперь при дефолтном состоянии плеера и при завершении проигрывания трека отображается время "00:00"

sobolevkir committed:
Доработана работа c избранным:
Теперь на экране плеера статус трека в избранном дополнительно не запрашивается. Вместо этого используется синхронизация и обновление списков в результатах поиска и истории через Flow из БД. Теперь в будущем можно будет реализовать возможность просмотра статуса избранного прямо в списке (например, сделать специальную иконку в элементах списков).

Кроме этого обратный порядок вывода списка избранных треков на экране Медиатеки теперь релизован с помощью дополнительного поля timestamp.

sobolevkir committed:
Merge remote-tracking branch 'origin/dev' into dev

sobolevkir commented:
> > > И что такое Database Inspector и как его реализовать в данном случае? Можно чуть подробнее, пожалуйста? 🙏
> > 
> > 
> > Внизу-слева есть раздел App Inspection. Если будут трудности, спроси в чате. Будет работать примерно так
> > 2024-08-02.23.22.09.mp4
> 
> Да, всё так и работает теперь! Но всё же я не понимаю, как мне сделать обновление списка в результатах поиска и в истории через этот flow, ведь треки, которые в результатах поиска находятся не из базы данных берутся, а через flow с сервера и поэтому этот список не обновляется при попадании трека в избранные. Может ты имел в виду, что нужно сделать апдейт списка (без сетевого запроса, но через интерактор) в методе onResume фрагмента? Мне нужно понимание, какое решение здесь наиболее удачным будет для обновления списка в результатах поиска и в истории...

В общем, я разобрался как работать с Flow. Теперь списки результата поиска и истории треков обновляются автоматически при добавлении трека в избранное на экране плеера и возврате к экрану поиска.
